### DISCUSSION TITLE
Weirdness with reactivity/signals?

### DISCUSSION THREAD

**Discussion by bdombro** [solidjs/solid#1627] [Q&A]
### Describe the bug

Hi, I'm experimenting with SolidJS and have noticed some strange patterns/possible bugs with reactivity. Wondering if y'all have pointers.

1. Child component doesn't react when I change the value of one of the props
2. The `For` component doesn't react when I update the `each` array
3. In contrast, using JSX.map on an array does react when the array changes.

Below is a preview of my experiment, and [here](https://stackblitz.com/edit/solidjs-templates-qpnmmr?file=src/App.tsx) is a Stackblitz interactive demo.

```typescript
/**
 * This component demonstrates weirdness in solidjs reactivity, especially
 * when passing reactive props.
 */
function ProductListComponent() {
  const [products, setProducts] = createSignal(productsInitial);
  const onIncrement = (id: number) =>
    setProducts((products) => {
      products.find((p) => p.id === id)!.id = id + 1;
      return [...products]; // return copy so SolidJS knows changed
    });
  // productsCopy: derived variables must be functions to be reactive. But note reactivity prob will fail if passed to a child as a prop.
  const productsCopy = () => [...products()];
  return (
	{/*
	Show the first product
	
	Solidjs doesn't catch when it changes. I'm not even sure how to
	fix. createStore suffers similar issues too.
	*/}
	<ProductDisplayComponent
		// signal variables must be called to get value. Notice that it
		// works on initial render but reactivity fails on change
		product={products()[0]}
		onIncrement={onIncrement}
	/>
	{/*
	List showing all products using 'For', the recommended way to map a list
	
	Solidjs doesn't catch that it changes. Again, I'm not sure how to fix.
	*/}
	<For each={products()}>
		{(p) => (
			<ProductDisplayComponent
				product={p}
				onIncrement={onIncrement}
			/>
		)}
	</For>
	{/*
	List showing all products using JSX.map</b>
	
	Solidjs catches changes! But not recommended and prob less
	performant.
	*/}
	{products().map((p) => (
	  <ProductDisplayComponent
		  product={p}
		  onIncrement={onIncrement}
	  />
	))}
	{/*
	List showing a copy of products using JSX.map

	Works as expected like.
	*/}
	{productsCopy().map((p) => (
	  <ProductDisplayComponent
		  product={p}
		  onIncrement={onIncrement}
	  />
	))}
  );
}
```

### Your Example Website or App

https://stackblitz.com/edit/solidjs-templates-qpnmmr?file=src/App.tsx

### Steps to Reproduce the Bug or Issue

Click the Increment buttons in the demo

### Expected behavior

I expected the children to react to changes to the signal. They are not.

### Platform

- OS: [e.g. macOS]
- Browser: Chrome
- Version: latest

**Comment by thetarnav** [comment]
`products.find((p) => p.id === id)!.id = id + 1`
You are mutating a property of a list item when `<For>` checks the referential equality of items to figure out what changed.
To fix this you can either:
1. destructure all modified objects instead of mutating
2. use a signal for the `id` property (nested in each item object) and set that signal instead - For won't detect that change, but any place that listens to the id signal will
3. use a store to have the same effect as in 2. but without manual wiring

**Comment by ryansolid** [comment]
Thanks @thetarnav. Moving this to discussions.

**Comment by bdombro** [comment]
@thetarnav I think you misunderstand the code -- the code DOES change reference. Please consider it in context -- the callback returned by setProducts returns a new array object by destructuring the source array with `[...products]`. The following line effectively makes a copy, which also creates a new reference, which I would expect SolidJS to catch.

So, it really does feel like a bug to me. Thoughts, @ryansolid ?

```typescript
const onIncrement = (id: number) =>
    setProducts((products) => {
      products.find((p) => p.id === id)!.id = id + 1;
      return [...products]; // return copy so SolidJS knows changed
    });
```

**Reply by ryansolid** [reply]
This isn't a bug you are mutating product. If you wanted it to truly be immutable you'd write:

```js
setProducts((products) => {
  return products.map(p => {
    if (p.id !== id) return p;
    return { ...p, id: p.id + 1 } // note clone the product as well
  }) // map clones the array
});
```
I still probably wouldn't recommend this over stores, but this is the pattern for updating items in lists in immutable data structures.

**Comment by bdombro** [comment]
Thanks, @ryansolid !
